RoPE models: add numerical sanity-check test for RoPE scaling

gante opened the pull request:
# What does this PR do?

Fixes #29765 

#29765 asks a pertinent question, where I had to look at the code to confirm the answer. This question should be checked automatically in a test instead -- confirm that RoPE scaling is working as intended.

gante committed:
add hard rope scaling test

gante commented on tests/models/llama/test_modeling_llama.py:
```diff
@@ -654,6 +654,325 @@ def decode_one_tokens(model, cur_token, input_pos, cache_position):
+
+        # The first sections of the Llama 2 paper. Input with >6k tokens, larger than the 4k model context window
+        tokenizer = LlamaTokenizer.from_pretrained("meta-llama/Llama-2-7b-hf", use_fast=True)
+        VERY_LONG_INPUT = '''
```
(see PR header on why we need this very long prompt)

gante commented on tests/models/llama/test_modeling_llama.py:
```diff
@@ -654,6 +654,325 @@ def decode_one_tokens(model, cur_token, input_pos, cache_position):
+        generate_kwargs = {"max_new_tokens": 40, "do_sample": False}
+        gen_out = model.generate(**model_inputs, **generate_kwargs)
+        decoded_text = tokenizer.decode(gen_out[0], skip_special_tokens=True)
+        self.assertTrue(decoded_text.endswith("The paper is about the development of Llama 2, a large language model (LLM) family, and the release of Llama 2-Chat, a fine-"))
```
This assertion would fail if #29765's suspicion [that we were applying the scaling factor twice] was true.

All these text terminations match the output on our original RoPE scaling release, meaning there was no regression since then :)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_29808). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

gante committed:
make fixup

amyeroberts reviewed (commented):
Thanks for adding this test! 

Just a few questions from my side: 
* Rather than generations, could we check the embedding values to see whether they've been rescaled as expected instead? 
* From the note it says it matches 'our initial rope scaling' - is this here for BC, or is it correct and just the initial feature? 
* Have the outputs been run and checked to compare the outputs if double scaling does happen?

gante commented:
> Rather than generations, could we check the embedding values to see whether they've been rescaled as expected instead?

Much better 👀 and doesn't need to be a slow test. Going to rework the PR

gante committed:
quick rope scaling tests

gante commented:
@amyeroberts deleted the previous slow test, and added a numerical test to the embeddings instead -- much much faster to test, and a more precise sanity check.

I've added a test on all RoPE-scaling compatible models. The test can't be abstracted into the mixin (for now), as there are a few variations while we are working on torch.compile/caching.

gante commented on tests/models/phi/test_modeling_phi.py:
```diff
@@ -360,6 +366,96 @@ def test_phi_sequence_classification_model_for_multi_label(self):
         self.assertEqual(result.logits.shape, (self.model_tester.batch_size, self.model_tester.num_labels))
 
+    @parameterized.expand([("linear",), ("dynamic",)])
+    def test_model_rope_scaling_from_config(self, scaling_type):
```
This one is not a new test, but rather a missing test. It is copy/paste from the same test in Falcon (Phi is mostly copied from Falcon)

gante commented on tests/models/falcon/test_modeling_falcon.py:
```diff
@@ -438,6 +443,65 @@ def test_model_rope_scaling(self, scaling_type):
         # The output should be different for long inputs
         self.assertFalse(torch.allclose(original_long_output, scaled_long_output, atol=1e-5))
 
+    def test_model_rope_scaling(self):
```
This test is the same on all models, with two variations:
- Llama passes `position_ids` to RoPE, as opposed to an integer depicting the sequence length
- GPTNeoX has `base=config.rotary_emb_base,`, all other models have `base=config.rope_theta`. This parameter is the same in GPTNeoX, but it has a different name

amyeroberts replied:
In this case, we can employ some `Copied from`s :)  

I don't think there's an easy way for the llama test - but for GPTNeox remapping with `with x->y` should work

amyeroberts reviewed (approved):
Looks great - super clear and easy to follow tests ❤️

gante committed:
add copy statements
